src/client/Client.cc

```diff
@@ -9056,7 +9056,7 @@ int Client::_readdir_cache_cb(dir_result_t *dirp, add_dirent_cb_t cb, void *p,
     }
 
     int idx = pd - dir->readdir_cache.begin();
-    if (dn->inode->is_dir()) {
+    if (dn->inode->is_dir() && cct->_conf->client_dirsize_rbytes) {
```

Comment:
The `client_dirsize_rbytes` option is changeable while running, after this will it still work ?

For example:

1, disable `client_dirsize_rbytes`
2, read the dir
3, stat the dir
4, enable `client_dirsize_rbytes`
5, stat the dir

Have you test whether the output in Step3 and Step5 are the same ? In theory it shouldn't be if the dir have non-zero size files.

Reply:
Could you add one test for this ?

Reply:
> Tested and it still works with the following program:
> 
> ```c
> 	assert(!ceph_create(&cmount, NAME));
> 	assert(!ceph_conf_set(cmount, "mon_host", MON));
> 	assert(!ceph_conf_set(cmount, "key", KEY));
> 	assert(!ceph_conf_set(cmount, "client_dirsize_rbytes", "false"));
> 	assert(!ceph_mount(cmount, TEST_ROOT));
> 	assert(!ceph_opendir(cmount, TEST_PATH, &dirp));
> 	assert((res = ceph_getdnames(cmount, dirp, buf, 4096)) > 0);
> 	printf("getdnames: %d, first %s\n", res, buf);
> 	assert(!ceph_closedir(cmount, dirp));
> 	assert(!ceph_stat(cmount, TEST_PATH, &st));
> 	printf("before: %d\n", st.st_size); // prints entries count
> 	assert(!ceph_conf_set(cmount, "client_dirsize_rbytes", "true"));
> 	assert(!ceph_stat(cmount, TEST_PATH, &st));
> 	printf("after: %d\n", st.st_size); // prints rbytes
> 	assert(!ceph_unmount(cmount));
> 	assert(!ceph_release(cmount));
> ```

In this I didn't see you created any file and write it under the `TEST_PATH`. Could you also share the output of your test ?

Reply:
> In this I didn't see you created any file and write it under the TEST_PATH.

This was tested against a real cluster with files populated beforehand in `TEST_PATH`.

I don't think I understand this test. Wouldn't stat also call `_getattr` by itself?

Reply:
Since you are removing the `CEPH_STAT_RSTAT` mask, and the MDS won't forward this to the auth MDS, so it may not update latest `rstat` to client. I am not sure when you enable that option and if do not force to get the `rstat` from auth MDS will the client just return the stale one if the caps needed are issued ? I didn't go through that code carefully yet, just wondering whether will this happen in some corner case ?

Reply:
Yeah, maybe I was wrong about this. I think we should explicitly make the `client_dirsize_rbytes` option to be not changeable in runtime ? Though it doesn't say it could be changed in runtime, but the common/options.cc code will allow it, please see https://github.com/ceph/ceph/commit/67524b14ebfbe21ebcffff7dc83f5e7a755ead80.

In kclient this is a mount option and is not changeable unless remounting it.

Reply:
In that case, let keep parity with kclient.

Reply:
@xdavidwu ping?

Reply:
I still don't think this will change behavior about changing `client_dirsize_rbytes` at runtime, but I will still add the restriction to be safe.